Importing pandas

In [1]:
import pandas as pd

Loading data

In [2]:
df = pd.read_csv("startup_funding2020.csv")

In [3]:
df.head()

Company/Brand Founded HeadQuarter              Sector  \
0    Aqgromalin    2019     Chennai            AgriTech   
1      Krayonnz    2019   Bangalore              EdTech   
2  PadCare Labs    2018        Pune  Hygiene management   
3         NCOME    2020   New Delhi              Escrow   
4    Gramophone    2016      Indore            AgriTech   

                                        What it does  \
0                       Cultivating Ideas for Profit   
1  An academy-guardian-scholar centric ecosystem ...   
2   Converting bio-hazardous waste to harmless waste   
3                       Escrow-as-a-service platform   
4  Gramophone is an AgTech platform enabling acce...   

                                            Founders  \
0                    Prasanna Manogaran, Bharani C L   
1                   Saurabh Dixit, Gurudutt Upadhyay   
2                                    Ajinkya Dhariya   
3                                      Ritesh Tiwari   
4  Ashish Rajan Singh, Harshit Gupta, Nishant Mah...   

                              Investor    Amount($)     Stage Column1  
0                      Angel investors    $200,000        NaN     NaN  
1                      GSF Accelerator    $100,000   Pre-seed     NaN  
2                       Venture Center  Undisclosed  Pre-seed     NaN  
3  Venture Catalysts, PointOne Capital    $400,000        NaN     NaN  
4  Siana Capital Management, Info Edge    $340,000        NaN     NaN

# Handling column "Founded"

In [4]:
df["Founded"].isnull().sum()

212

We have found 212 empty values. we can replaced them by either by mode or mean of the column. our method is going to copy the column and work for calculatin mean and mode.

In [5]:
my_record = df.Founded # copying the column

In [6]:
my_record[482]

'-'

As shown in the above the recorder has "-" and "nan" values. They have to be dropped inorder to find mean and mode.

In [7]:
# searching for the index of the value "-" from the column
for i in range(df.shape[0]):
    if my_record[i] == "-":
        print(i)

482


In [8]:
my_record = my_record.drop(482)

In [1]:
my_record = my_record.dropna()

NameError: name 'my_record' is not defined

In [10]:
my_record.unique()

array(['2019', '2018', '2020', '2016', '2008', '2015', '2017', '2014',
       '1998', '2007', '2011', '1982', '2013', '2009', '2012', '1995',
       '2010', '2006', '1978', '1999', '1994', '2005', '1973', '2002',
       '2004', '2001'], dtype=object)

Now our data is free from error for our calculation. It need only to convert it to integer

In [11]:
my_record = my_record.astype(int) # converting to int data type

In [12]:
my_record.info()

<class 'pandas.core.series.Series'>
Int64Index: 842 entries, 0 to 1054
Series name: Founded
Non-Null Count  Dtype
--------------  -----
842 non-null    int32
dtypes: int32(1)
memory usage: 9.9 KB


In [13]:
my_record.mode()

0    2015
Name: Founded, dtype: int32

The Mode is 2015

In [14]:
my_record.mean()

2015.3634204275534

The average is 2015. So we can use 2015 as an imputer for missed value for column founded.

In [15]:
df.Founded.unique()

array(['2019', '2018', '2020', '2016', '2008', '2015', '2017', '2014',
       '1998', '2007', '2011', '1982', '2013', '2009', '2012', '1995',
       '2010', '2006', '1978', nan, '1999', '1994', '2005', '1973', '-',
       '2002', '2004', '2001'], dtype=object)

In [16]:
df.Founded = df.Founded.replace("-",2015) # repalcing "-" by 2015

In [17]:
df.Founded.unique()

array(['2019', '2018', '2020', '2016', '2008', '2015', '2017', '2014',
       '1998', '2007', '2011', '1982', '2013', '2009', '2012', '1995',
       '2010', '2006', '1978', nan, '1999', '1994', '2005', '1973', 2015,
       '2002', '2004', '2001'], dtype=object)

In [18]:
df.Founded = df.Founded.fillna(2015) # replacing NaN 

In [19]:
df.Founded.unique()

array(['2019', '2018', '2020', '2016', '2008', '2015', '2017', '2014',
       '1998', '2007', '2011', '1982', '2013', '2009', '2012', '1995',
       '2010', '2006', '1978', 2015, '1999', '1994', '2005', '1973',
       '2002', '2004', '2001'], dtype=object)

In [20]:
df.Founded.isnull().any()

False

# Handling column "Company/Brand"

In [21]:
df.rename(columns = { "Company/Brand" : "Company Name"}, inplace=True) # rename the column

In [22]:
df.columns

Index(['Company Name', 'Founded', 'HeadQuarter', 'Sector', 'What it does',
       'Founders', 'Investor', 'Amount($)', 'Stage', 'Column1'],
      dtype='object')

# Handling column "Head Quarter"

In order to make columns be same with startup_funding2018. The name of countries should be renamed as "City"

In [23]:
df.rename(columns = { "HeadQuarter" : "City"}, inplace=True) # rename the column

In [24]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'What it does', 'Founders',
       'Investor', 'Amount($)', 'Stage', 'Column1'],
      dtype='object')

In [25]:
df.Sector.isnull().sum() # checking if there is null value

13

The column has 13 empty values. we are going to substitute them with the most frequent value. Before that the indeex of empty value have to be determined.

In [26]:
my_record2 = df.Sector.isnull() # copying the column

In [27]:
mode_val = df.Sector.mode() # finding frequented value

In [28]:
for i in range(len(my_record2)):
    if my_record2[i] == True: # searching for empty/nan value
        df.Sector[i] = mode_val

In [29]:
df.Sector.isnull().any()

False

In [30]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'What it does', 'Founders',
       'Investor', 'Amount($)', 'Stage', 'Column1'],
      dtype='object')

# Handling column "What it does"

In [31]:
df["What it does"].isnull()

0       False
1       False
2       False
3       False
4       False
        ...  
1050    False
1051    False
1052    False
1053    False
1054    False
Name: What it does, Length: 1055, dtype: bool

For the sake matching columns be same as in other excels. we give similar column name for company detail

In [32]:
df.rename(columns = { "What it does" : "About Company"}, inplace=True) # rename the column

In [33]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount($)', 'Stage', 'Column1'],
      dtype='object')

# Handling column "Founders"

In [34]:
df.Founders.isnull().sum()

12

The column has 12 null values, first we count them and replace with a flag called "unknown"

In [35]:
my_record3 = df.Founders.isnull()

In [36]:
for i in range(len(my_record3)):
    if my_record3[i] == True:
        df.Founders[i] = "unkown"

In [37]:
df.Founders.isnull().sum()

0

In [38]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount($)', 'Stage', 'Column1'],
      dtype='object')

# Handling column "Investor"

In [39]:
df.Investor.isnull().sum()

38

The column has 32 null values, first we count them and replace with a flag called "unknown"

In [40]:
my_record4 = df.Investor.isnull()

In [41]:
for i in range(len(my_record4)):
    if my_record4[i] == True:
        df.Investor[i] = "unknown"

In [42]:
df.Investor.isnull().sum()

0

In [43]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount($)', 'Stage', 'Column1'],
      dtype='object')

# Handling column "Amount"

In [44]:
df["Amount($)"].unique()

array(['$200,000 ', '$100,000 ', 'Undisclosed', '$400,000 ', '$340,000 ',
       '$600,000 ', '$45,000,000 ', '$1,000,000 ', '$2,000,000 ',
       '$1,200,000 ', '$660,000,000 ', '$120,000 ', '$7,500,000 ',
       '$5,000,000 ', '$500,000 ', '$3,000,000 ', '$10,000,000 ',
       '$145,000,000 ', '$100,000,000 ', nan, '$21,000,000 ',
       '$4,000,000 ', '$20,000,000 ', '$560,000 ', '$275,000 ',
       '$4,500,000 ', '$15,000,000 ', '$390,000,000 ', '$7,000,000 ',
       '$5,100,000 ', '$700,000,000 ', '$2,300,000 ', '$700,000 ',
       '$19,000,000 ', '$9,000,000 ', '$40,000,000 ', '$750,000 ',
       '$1,500,000 ', '$7,800,000 ', '$50,000,000 ', '$80,000,000 ',
       '$30,000,000 ', '$1,700,000 ', '$2,500,000 ', '$40,000 ',
       '$33,000,000 ', '$35,000,000 ', '$300,000 ', '$25,000,000 ',
       '$3,500,000 ', '$200,000,000 ', '$6,000,000 ', '$1,300,000 ',
       '$4,100,000 ', '$575,000 ', '$800,000 ', '$28,000,000 ',
       '$18,000,000 ', '$3,200,000 ', '$900,000 ', '$250,000 '


This column has nan and undisclosed values. We are going to substitue them by mean amount.


In [45]:
my_record5 = df["Amount($)"] # copying the column

In [46]:
# replacing values

my_record5 = my_record5.apply(lambda x: str(x).replace(",",""))
my_record5 = my_record5.apply(lambda x: str(x).replace("Undisclosed","0"))
my_record5 = my_record5.apply(lambda x: str(x).replace("$",""))
my_record5 = my_record5.apply(lambda x: str(x).replace("Undislosed","0"))
my_record5 = my_record5.apply(lambda x: str(x).replace("Undiclsosed","0"))
my_record5 = my_record5.apply(lambda x: str(x).replace("nan","0"))


In [47]:
# eliminating space

state = my_record5.str.endswith(" ")

for i in range(len(my_record5)):
    if state[i]:
        my_record5[i] = my_record5[i][:-1]

In [48]:
# there are two columns having two values

val = {}
for i in range(len(my_record5)):
    if " " in my_record5[i]:
        val[i] = my_record5[i]
val

{465: '887000 23000000', 472: '800000000 to 850000000'}

In [49]:
# a. replacing values with their mean

val2 = val[465].split(" ")
val2

a = float(val2[0])
b = float(val2[1])

mean = (a + b) / 2
my_record5[465] = mean

mean = (a + b) / 2
my_record5[465] = mean

In [50]:
# b. replacing values with their mean

val3 = val[472].split(" to ")
val3

c = float(val2[0])
d = float(val2[1])

mean2 = (c + d) / 2
my_record5[472] = mean2

In [51]:
my_record5 = my_record5.astype(float) # converting to float

In [52]:
my_record5 = my_record5.replace(0.0 ,mean2) # replacing by mean

In [53]:
df["Amount($)"] = my_record5 # substituting the columns

In [54]:
df["Amount($)"]

0         200000.0
1         100000.0
2       11943500.0
3         400000.0
4         340000.0
           ...    
1050     1500000.0
1051    13200000.0
1052     8000000.0
1053     8043000.0
1054     9000000.0
Name: Amount($), Length: 1055, dtype: float64

In [55]:
df = df.rename(columns={"Amount($)": "Amount"})

# Handling column "Stage"

In [56]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount', 'Stage', 'Column1'],
      dtype='object')

In [57]:
df.Stage.unique()

array([nan, 'Pre-seed', 'Seed', 'Pre-series A', 'Pre-series', 'Series C',
       'Series A', 'Series B', 'Debt', 'Pre-series C', 'Pre-series B',
       'Series E', 'Bridge', 'Series D', 'Series B2', 'Series F',
       'Pre- series A', 'Edge', 'Series H', 'Pre-Series B', 'Seed A',
       'Series A-1', 'Seed Funding', 'Pre-Seed', 'Seed round',
       'Pre-seed Round', 'Seed Round & Series A', 'Pre Series A',
       'Pre seed Round', 'Angel Round', 'Pre series A1', 'Series E2',
       'Pre series A', 'Seed Round', 'Bridge Round', 'Pre seed round',
       'Pre series B', 'Pre series C', 'Seed Investment', 'Series D1',
       'Mid series', 'Series C, D', 'Seed funding'], dtype=object)

In [58]:
df.Stage = df.Stage.fillna(df.Stage.mode()[0])

In [59]:
df.Stage.unique()

array(['Series A', 'Pre-seed', 'Seed', 'Pre-series A', 'Pre-series',
       'Series C', 'Series B', 'Debt', 'Pre-series C', 'Pre-series B',
       'Series E', 'Bridge', 'Series D', 'Series B2', 'Series F',
       'Pre- series A', 'Edge', 'Series H', 'Pre-Series B', 'Seed A',
       'Series A-1', 'Seed Funding', 'Pre-Seed', 'Seed round',
       'Pre-seed Round', 'Seed Round & Series A', 'Pre Series A',
       'Pre seed Round', 'Angel Round', 'Pre series A1', 'Series E2',
       'Pre series A', 'Seed Round', 'Bridge Round', 'Pre seed round',
       'Pre series B', 'Pre series C', 'Seed Investment', 'Series D1',
       'Mid series', 'Series C, D', 'Seed funding'], dtype=object)

# Handling column "Column1"

In [60]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount', 'Stage', 'Column1'],
      dtype='object')

In [61]:
df = df.drop(columns = "Column1")

In [62]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount', 'Stage'],
      dtype='object')

# removing duplicate values

In [63]:
df = df.loc[df.astype(str).drop_duplicates().index]

In [64]:
df

Company Name Founded       City              Sector  \
0       Aqgromalin    2019    Chennai            AgriTech   
1         Krayonnz    2019  Bangalore              EdTech   
2     PadCare Labs    2018       Pune  Hygiene management   
3            NCOME    2020  New Delhi              Escrow   
4       Gramophone    2016     Indore            AgriTech   
...            ...     ...        ...                 ...   
1050  Leverage Edu    2015      Delhi              Edtech   
1051         EpiFi    2015        NaN             Fintech   
1052       Purplle    2012     Mumbai           Cosmetics   
1053        Shuttl    2015      Delhi           Transport   
1054         Pando    2017    Chennai            Logitech   

                                          About Company  \
0                          Cultivating Ideas for Profit   
1     An academy-guardian-scholar centric ecosystem ...   
2      Converting bio-hazardous waste to harmless waste   
3                          Escrow-as-a-service platform   
4     Gramophone is an AgTech platform enabling acce...   
...                                                 ...   
1050  AI enabled marketplace that provides career gu...   
1051  It offers customers with a single interface fo...   
1052         Online makeup and beauty products retailer   
1053                    App based bus aggregator serice   
1054            Networked logistics management software   

                                               Founders  \
0                       Prasanna Manogaran, Bharani C L   
1                      Saurabh Dixit, Gurudutt Upadhyay   
2                                       Ajinkya Dhariya   
3                                         Ritesh Tiwari   
4     Ashish Rajan Singh, Harshit Gupta, Nishant Mah...   
...                                                 ...   
1050                                  Akshay Chaturvedi   
1051                    Sujith Narayanan, Sumit Gwalani   
1052                          Manish Taneja, Rahul Dash   
1053                      Amit Singh, Deepanshu Malviya   
1054                     Jayakrishnan, Abhijeet Manohar   

                                   Investor      Amount       Stage  
0                           Angel investors    200000.0    Series A  
1                           GSF Accelerator    100000.0    Pre-seed  
2                            Venture Center  11943500.0    Pre-seed  
3       Venture Catalysts, PointOne Capital    400000.0    Series A  
4       Siana Capital Management, Info Edge    340000.0    Series A  
...                                     ...         ...         ...  
1050  DSG Consumer Partners, Blume Ventures   1500000.0    Series A  
1051          Sequoia India, Ribbit Capital  13200000.0  Seed Round  
1052                             Verlinvest   8000000.0    Series A  
1053             SIG Global India Fund LLP.   8043000.0    Series C  
1054                      Chiratae Ventures   9000000.0    Series A  

[1052 rows x 9 columns]

# Renaming to standard

In [65]:
df.columns

Index(['Company Name', 'Founded', 'City', 'Sector', 'About Company',
       'Founders', 'Investor', 'Amount', 'Stage'],
      dtype='object')

In [69]:
df = df.rename(columns={"Company Name": "Company/Brand", "About Company": "What it does", 'Amount' : "Amount($)"})

In [70]:
df.columns

Index(['Company/Brand', 'Founded', 'City', 'Sector', 'What it does',
       'Founders', 'Investor', 'Amount($)', 'Stage'],
      dtype='object')

In [71]:
file_name = "cleaned 2020.xlsx"

df.to_excel(file_name)

## Finally we have removed, in excel side, Column1 